src/query/storage/promremote/options.go

```diff
@@ -0,0 +1,133 @@
+)
+
+// NewOptions constructs Options based on the given config.
+func NewOptions(cfg *config.PrometheusRemoteBackendConfiguration, scope tally.Scope) (Options, error) {
+	err := validateBackendConfiguration(cfg)
+	if err != nil {
+		return Options{}, err
+	}
+	endpoints := make([]EndpointOptions, len(cfg.Endpoints))
+
+	for i, endpoint := range cfg.Endpoints {
+		endpoints[i] = EndpointOptions{
+			name:    endpoint.Name,
+			address: endpoint.Address,
+		}
+		if endpoint.StoragePolicy != nil {
+			endpoints[i].resolution = endpoint.StoragePolicy.Resolution
+			endpoints[i].retention = endpoint.StoragePolicy.Retention
+		}
+	}
+	clientOpts := xhttp.DefaultHTTPClientOptions()
+	if cfg.RequestTimeout != nil {
+		clientOpts.RequestTimeout = *cfg.RequestTimeout
+	}
+	if cfg.ConnectTimeout != nil {
+		clientOpts.ConnectTimeout = *cfg.ConnectTimeout
+	}
+	if cfg.KeepAlive != nil {
+		clientOpts.KeepAlive = *cfg.KeepAlive
+	}
+	if cfg.IdleConnTimeout != nil {
+		clientOpts.IdleConnTimeout = *cfg.IdleConnTimeout
+	}
+	if cfg.MaxIdleConns != nil {
+		clientOpts.MaxIdleConns = *cfg.MaxIdleConns
+	}
+
+	clientOpts.DisableCompression = true // Already snappy compressed.
+
+	return Options{
+		endpoints:   endpoints,
+		httpOptions: clientOpts,
+		scope:       scope,
+	}, nil
+}
+
+func validateBackendConfiguration(cfg *config.PrometheusRemoteBackendConfiguration) error {
+	if cfg == nil {
+		return fmt.Errorf("prometheusRemoteBackend configuration is required")
+	}
+	if len(cfg.Endpoints) == 0 {
+		return fmt.Errorf(
+			"at least one endpoint must be configured when using %s backend type",
+			config.PromRemoteStorageType,
+		)
+	}
+	if cfg.MaxIdleConns != nil && *cfg.MaxIdleConns < 0 {
+		return errors.New("maxIdleConns can't be negative")
+	}
+	if cfg.KeepAlive != nil && *cfg.KeepAlive < 0 {
+		return errors.New("keepAlive can't be negative")
+	}
+	if cfg.IdleConnTimeout != nil && *cfg.IdleConnTimeout < 0 {
+		return errors.New("idleConnTimeout can't be negative")
+	}
+	if cfg.RequestTimeout != nil && *cfg.RequestTimeout < 0 {
+		return errors.New("requestTimeout can't be negative")
+	}
+	if cfg.ConnectTimeout != nil && *cfg.ConnectTimeout < 0 {
+		return errors.New("connectTimeout can't be negative")
+	}
+
+	seenNames := map[string]struct{}{}
+	for _, endpoint := range cfg.Endpoints {
+		if err := validateEndpointConfiguration(endpoint); err != nil {
+			return err
+		}
+		if _, ok := seenNames[endpoint.Name]; ok {
+			return fmt.Errorf("endpoint name %s is not unique, ensure all endpoint names are unique", endpoint.Name)
+		}
+		seenNames[endpoint.Name] = struct{}{}
+	}
+	return nil
+}
+
+func validateEndpointConfiguration(endpoint config.PrometheusRemoteBackendEndpointConfiguration) error {
+	if endpoint.StoragePolicy != nil {
+		if endpoint.StoragePolicy.Resolution < 0 {
+			return errors.New("endpoint resolution can't be negative")
+		}
+		if endpoint.StoragePolicy.Retention < 0 {
+			return errors.New("endpoint retention can't be negative")
+		}
```

Comment:
Perhaps now that `storagePolicy` itself is optional, you should check that resolution/retention are strictly positive?